### PR TITLE
[BEAM-11805] Replace user-agent for spanner

### PR DESCRIPTION
**Pull request by allenpradeep**
This patch contains the workaround to replace the user-agent for spanner.
R: @TheNeuralBit  @nielm 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by allenpradeep** [comment]
@TheNeuralBit  - This patch has the following checkstyle violation. Can you help with this?
[ant:checkstyle] [ERROR] /usr/local/google/home/alxavier/spannerio/apbeam/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerAccessor.java:35: You are using raw guava, please use vendored guava classes. [ForbidNonVendoredGuava]

**Comment by nielm** [comment]
> @TheNeuralBit - This patch has the following checkstyle violation. Can you help with this?
> [ant:checkstyle] [ERROR] /usr/local/google/home/alxavier/spannerio/apbeam/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerAccessor.java:35: You are using raw guava, please use vendored guava classes. [ForbidNonVendoredGuava]

Replace the import 
com.google.common.util.concurrent.ThreadFactoryBuilder
With
org.apache.beam.vendor.guava.v26_0_jre.com.google.common.util.concurrent.ThreadFactoryBuilder

**Review by thiagotnunes — APPROVED** [review]

**Comment by allenpradeep** [comment]
Run Java PreCommit

**Comment by allenpradeep** [comment]
@nielm  @thiagotnunes  - Thanks

Please take a look.
I have fixed the checkstyle issues and added the support for checking emulator connection.

**Review by thiagotnunes — APPROVED** [review]

**Review by TheNeuralBit — CHANGES_REQUESTED** [review]
@allenpradeep could you also cherry-pick the the version update commits from #13765? They were rolled back in #13792. I'd like to apply both changes at once to make sure they pass CI before merging.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerAccessor.java
Comment (suztomo) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerAccessor.java:202]: Is this a link to a bug to be fixed (I see it's been merged already)?

If not, would you also add the link to the bug?

Comment (thiagotnunes) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerAccessor.java:202]: - https://github.com/googleapis/java-spanner/pull/747 is the PR where the bug was introduced.
- https://github.com/googleapis/java-spanner/issues/880 is the issue.
- https://github.com/googleapis/java-spanner/pull/871 is the PR for the bug fix.

Comment (allenpradeep) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerAccessor.java]: Added https://github.com/googleapis/java-spanner/pull/871 in the comment.

**Comment by allenpradeep** [comment]
Run Java PreCommit

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerAccessor.java
Comment (TheNeuralBit) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerAccessor.java:198]: Why add an `else` here? The previous logic called `setHeaderProvider` unconditionally.

**Comment by allenpradeep** [comment]
> @allenpradeep could you also cherry-pick the the version update commits from #13765? They were rolled back in #13792. I'd like to apply both changes at once to make sure they pass CI before merging.

There were some spanner test issues with the previous commit. I'll cherry pick #13765 after the integration tests passes and resend.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerAccessor.java
Comment (suztomo) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerAccessor.java]: Thank you.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerAccessor.java
Comment (allenpradeep) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerAccessor.java:198]: This is to enable a fail fast if the emulator is not running, but the user has set the emulator host environment variable.

As per [function-checkEmulatorConnection](https://github.com/googleapis/java-spanner/blob/master/google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpc.java#L497-L539) [line 502](https://github.com/googleapis/java-spanner/blob/master/google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpc.java#L502), the fail fast link works only if the channelprovider is set to null. This workaround only sets the channel provider only when we are not going to connect to emulator.

cc @thiagotnunes

**Comment by allenpradeep** [comment]
> @allenpradeep could you also cherry-pick the the version update commits from #13765? They were rolled back in #13792. I'd like to apply both changes at once to make sure they pass CI before merging.

@TheNeuralBit  - cherry-picked the the 2 commits from #13765.
Waiting for the tests to complete.

**Comment by allenpradeep** [comment]
Run Java PreCommit

**Comment by allenpradeep** [comment]
Run Java PreCommit

**Comment by TheNeuralBit** [comment]
@allenpradeep  The merge conflict looks to be due to https://github.com/apache/beam/pull/13804, which upgraded guava. I don't think it's an actual conflict, it's just that 007c0ab is upgrading gRPC, which is on an adjacent line. The resolution should be to just make sure we upgrade both.

**Comment by allenpradeep** [comment]
> @allenpradeep The merge conflict looks to be due to #13804, which upgraded guava. I don't think it's an actual conflict, it's just that [007c0ab](https://github.com/apache/beam/commit/007c0ab955b73962c5188412b446f2bec0408cc7) is upgrading gRPC, which is on an adjacent line. The resolution should be to just make sure we upgrade both.

@TheNeuralBit 
Done. PTAL

**Comment by allenpradeep** [comment]
Run Java PostCommit

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerAccessor.java
Comment (TheNeuralBit) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerAccessor.java:198]: Ah I see, thank you.

**Review by TheNeuralBit — APPROVED** [review]
Thank you @allenpradeep, I will merge when CI is green.

**Comment by TheNeuralBit** [comment]
Failure is a flake in FhirIOReadIT. I'll go ahead and merge.

**Pull request merged**
